Add an addAllUserCourses option and fix a couple of bugs

RedEtherbloom opened the pull request:
The base of this PR a new function **addAllUserCourses** that parses the users courses out of the login soup, so you don't have to lookup every ref_id if you want to download all your courses.

During testing I uncovered multiple inconsistencies between courses, that clashed with assumptions of the Downloader Code and crashed the program. I found three of these. I'll open separate issues for these later, and crosslink them.
I tried to make the parsing an these places a bit more dynamic and robust to avoid similar problems in the future.

Hope I didn't touch you code too much, I tried to give it my best and make the PR as clean as possible for me.
My commit history should be pretty verbose if you're unsure about what I intended with some functions.

Hope my PR will be helpful :)
Oh and if you think that I did a good job, you can add the **hacktoberfest-accepted** label to my PR to help me with my [Hacktoberfest](https://hacktoberfest.digitalocean.com/) progress

jhelgert committed:
Dev (#20)

* Fix parsing of file properties for missing filename extension

jhelgert committed:
Fixed version number

Commit:
Added stub for extractUserCourses

I decided to split the quering of the subscribed users courses into
extractUserCourses and later addAllUserCourses.

I chose this approach so to make this functionality:
a) more readable
b) more flexible
This way, if a user of this library, ever needs the user information for
other purposes they can call this function, instead of having to extract
this information manually out of the self.courses.

Commit:
Wrote most of the extractUserCourses code

Decdided to split the commit here, I'll have to write a function to
extract the ref_id.

Commit:
Wrote extractIDfromURL stub and Docstring

Commit:
Wrote extractIdFromUrl code and fixed camel case

Since there shouldn't be two ref_ids in the same URL the code is simply
run RegEx with group, check if there were matches, if yes, it's valid,
so return the only matched group.

Commit:
Completed extractUserCourses code and updated doc

I tried to model the courses list that gets returned after the
self.courses of the class, with the url parameter beign replaced by
ref_id.

Commit:
Wrote code for addAllUserCourses

addAllUserCourses uses the ref_ids that were extracted from the ilias
urls and adds them via addCourses.

The reason why I don't simply add name and url myself to the list is to
avoid any mistakes on my side, and have only one mechanism for adding
courses in the code. This keeps the code more flexible and easier to
change.
I can also not be fully sure that the url parsed from the login_soup
won't have any random, unneeded that in it. So this keeps the
course urls consistent as well.

Code works for all courses I've tested so far, except one of cause...
Course 342184 makes problems, because it's inconsistent with other
courses formats. Problem will be addressed in next commit.

Commit:
Removed forgotten debug print

Commit:
Reworked addCourse

addCourse tried to find the course name statically. Some courses use a
slightly different breadcrumb pattern, not sure how intended this is.

The course name is now found dynamically via the <a> that contains it
and points back onto the course page.

This fixes the problem mentioned in 9867ec82248dec929e2030ad3a87b481b6387717
and makes the parsing more robust.

Commit:
Added cleanCourseName function

Decoupled cleaning the course name, so that it can be used in
extractUserCourses as well, instead of repasting the regex.

Commit:
Replaced regex with cleanCourseName

Commit:
Added iliasid to courses dict and replaced ref_id

The ref_id got replaced with iliasid to be more consistent with the rest
of the code.

The iliasid got added to the dictionary to avoid having to parse it out
if it'll would ever be needed for anything. It also doesn't add much
redundant information imho.

Commit:
Renamed ref_id in extractIdFromUrl

Renamed ref_id to iliasid to match the rest of the class.

Commit:
Reworked video filename system

The ilias is inconsistent again and not every video has caption.
So we have to work with the basename in that case instead.

Taking the basename and just removing the last extension seems like the
best option for now.

This system won't be perfect. People sometimes like to use . as a word
seperator in filenames. If a file had no extension and used dots as word
seperators the last word would get swallowed.
Other problem would be if the file base name had multiple file
extensions(If the file eg got converte and they just added an
addiotional extension to the filename).
People are incosistent, so there isn't a perfect solution.
I'd call my current approach a good compromise between removing too much
or too little.

Commit:
Fixed exception if _parseFileProperties

Ilias has more incosistencies. Some file il_ItemProperty attributes
contain just the file extenson, nothing more.

I could technically have parsed that info from the file info page.
But then we'd have two different code paths and the ilias request would
have slowed down the code even further.

Commit:
Added addAllUserCourses example to README.md

GityUpNow commented:
PR should fix #22 #23 #24

jhelgert commented:
At first sight it seems fine. I'll have a detailed look tomorrow. :)

jhelgert commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -127,6 +173,44 @@ def addCourses(self, *iliasids):
+		return num_added
+
+
+	def cleanCourseName(self, course_name):
+		"""
+		Cleans up a course name to remove unwanted data.
+
+		:param      course_name:  raw course_name(from login_soup, or request)
+		:type       course_name:  str
+
+		:returns:   clean course name
+		:rtype:     str
+		"""
+
+		return re.sub(r"\[.*\] ", "", course_name).strip()
```
One-liner and only used once. Should be removed. Using it with a comment should be sufficient.

GityUpNow replied:
It's actually used in two places,  addAllCourses(or whatever it's name will be in the end) and addCourse.
Didn't want to paste the RegEx in two places, so I went for the helper method.
I also did these because of the design philosophy in mind, look at my newest answer for that one.

jhelgert commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -163,21 +247,28 @@ def _determineItemType(self, url):
 		if len(p[0].text.split()) > 1:
 			file_ending = ""
 		else:
 			file_ending = "." + p[0].get_text().split()[0]
-		file_size_tmp = p[1].get_text().replace(".","").replace(",", ".").split()
-		file_size = float(file_size_tmp[0])
-		if file_size_tmp[1] == "KB":
-			file_size *= 1e-3
-		elif file_size_tmp[1].lower() == "bytes":
-			file_size *= 1e-6
+
+		if len(p) > 1:
+			file_size_tmp = p[1].get_text().replace(".","").replace(",", ".").split()
+			file_size = float(file_size_tmp[0])
+			if file_size_tmp[1] == "KB":
+				file_size *= 1e-3
+			elif file_size_tmp[1].lower() == "bytes":
+				file_size *= 1e-6
+		else:
+			file_size = math.nan
+
 		p = [i for i in p if "Version" not in i.text]
 		if len(p) > 2:
 			file_mod_date = parsedate(self.translate_date(p[2].get_text()), dayfirst=True)
 		else:
 			file_mod_date = datetime.fromisoformat('2000-01-01')
+
 		return file_ending, file_size, file_mod_date
```
Never been really proud about this method, it's by far the most ugliest one inside the whole class. 😅 So it seems to be a good idea to add a few comments regarding Ilias' quirks and refactor this one a bit in the near future. 

Are you willed to help?

GityUpNow replied:
I'd be more than willing to help, but I think I'll two to three weeks before that.
I still have to send three PRs for Hacktoberfest and got a lot of other stuff to do first.

GityUpNow replied:
The function was probably a prototype to get things off the ground by you?
I see nothing wrong with testing things things first and redoing them at some later point :shrug:

GityUpNow replied:
If I try to contribute tests to this project at some point I'd think this would be a good method to start.

jhelgert replied:
> I'd be more than willing to help, but I think I'll two to three weeks before that.

Don't worry, that's not urgent.

> The function was probably a prototype to get things off the ground by you?

Exactly, even though I don't see a more elegant way to do it atm. 🤔

> If I try to contribute tests to this project at some point I'd think this would be a good method to start.

👍

jhelgert commented on README.md:
```diff
@@ -53,6 +53,19 @@ m.addCourse(965389)   # BE 511 Business Economics II
+m = IliasDownloaderUniMA()
+m.setParam('download_path', '/Users/jonathan/Desktop/')
+m.login('jhelgert', 'my_password')
+m.addAllUserCourses()
```
```suggestion
m.addAllCurrentSemesterCourses()
```

GityUpNow replied:
See my comment to your first addAllUserCourses proposal.
TL;DR I'd pick addAllSemesterCourses, the method name is even with addAllSemesterCourses already a bit too long imo, but borderline okayish I'd say.

jhelgert commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -127,6 +173,44 @@ def addCourses(self, *iliasids):
 			self.addCourse(iliasid)
 
 
+	def addAllUserCourses(self, excluded_ids=[]):
```
```suggestion
	def addAllCurrentSemesterCourses(self):
```

GityUpNow replied:
I don't fully get why you don't want the excluded_ids parameter.
It'd doesn't add that much complexity, but makes the method much more flexible.
First use case that I could think off: Some prof inserted another inconsistent, haunted structure in their course that breaks the parser. Course could be temporarily disabled this way til someone fixed it.

jhelgert replied:
> I don't fully get why you don't want the excluded_ids parameter.

I wasn't sure if you were aware of the problems withs `extractUserCourses()` and indented the user to exclude all unwanted ids. Now I see your point, so let's keep the `exlucuded_ids`. But `None` as default value is more consistent IMHO.

jhelgert reviewed (commented):
Thanks for the PR, really appreciated! Only a few changes here and there and further testing needed.

jhelgert commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -198,8 +289,19 @@ def scanFolder(self, course_name, url_to_scan):
 		for v in videos:
-			el_url = urljoin(self.base_url, v.find('source')['src'])
-			el_name = v.find('div', {'class': 'ilc_media_caption_MediaCaption'}).get_text()
+			v_url = v.find('source')['src']
```
This breaks if the media inside `ilc_media_caption_MediaCaption` isn't a video. Example:

<img width="573" alt="Bildschirmfoto 2020-10-23 um 12 19 07" src="https://user-images.githubusercontent.com/8471326/96993168-37455600-152b-11eb-9803-99a974bbd372.png">

```suggestion
			try:
				v_url = v.find('source')['src']
			except:
				continue
```

GityUpNow replied:
Didn't think of this problem. Looks like a solid change, I'll commit it.

Sidenote: I think we have to implement or import a bw sync&share downloader at some point. 
Some profs upload some files, only saw videos thus far, there and link them back, instead of uploading them directly :roll_eyes:

jhelgert commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -100,6 +120,29 @@ def login(self, login_id, login_pw):
 			raise ConnectionError("Couldn't log into ILIAS. Make sure your provided uni-id and the password are correct.")
 
 
+	def extractUserCourses(self):
+		"""
+		Extracts the users subscribed courses from the login_soup
+
+		:returns:   List of dicts, with keys name, url and iliasid
+		:rtype:     list
+		"""
+
+		courses = []
+
+		containers = self.login_soup.select("div.il_ContainerListItem")
+		for container in containers:
+			title_elem = container.select("a.il_ContainerItemTitle")[0]
+
+			course_name = self.cleanCourseName(title_elem.string)
+			iliasid = self.extractIdFromUrl(title_elem["href"])
+
+			url = self.createIliasUrl(iliasid)
+			courses += [{"name": course_name, "url": url, "iliasid": iliasid}]
+
+		return courses
```
```suggestion
	def addAllCurrentSemesterCourses(self):
		"""
		Adds all current semester courses.
		"""

		items = self.login_soup.find_all("a", "il_ContainerItemTitle")

		# Find the current semester (should be the first entry
		# matching the regex pattern)
		for i in items:
			match = re.search(r"([A-Z]{3} \d{4})", i.text)
			if match:
				current_semester = match.group(0)
				break
		# add the current semester's courses
		for course in items:
			if current_semester in course.text:
				url = course['href']
				ilias_id = int(re.search(r"ref_id=(\d+)", url).group(1))
				self.addCourse(ilias_id)
```

GityUpNow replied:
See my comments to your proposals for extractUserCourses and addAllCourses above, plus my design idea below.
I don't like this change tbh. It removes the excluded_ids, works only for the current semester and will make things a lot harder to test.
I'd prefer a flexible over a rigid design.

jhelgert commented:
PS: Could you please change the pull requests' branch to 'dev' instead of 'master'? While we're at it, I'll then clean up a few other parts of my old code - especially `scanFolder()`.

GityUpNow commented:
Sure, I can try to do that later. Hope this won't create too many merge conflicts.

So, about the design philosophy for my code that I mentioned:
Ilias is imho often a pretty chaotic system, to a good part because of inconsistencies between courses. This meant that I wanted to make the code as robust as I can.
This lead to the next conclusion: Maybe I want to contribute some tests to this project at some point.
So I wanted to make the code easily testable.

This influenced how I wrote my code, split tasks into logical smaller methods, so that I'll be easily able to test them later.
An example for this would be extractUserCourses and addAllUserCourses. The latter calls the former to get it's data, so I'll be able to just mock out the former and return a test-list, without having to write a login_soup for every test scenario.

Since you have some smaller functions like createIliasUrl in your code, I thought my code design should fit pretty good in there.
I wanted to also make my methods a bit future proof. Sub tasks that might be useful in the future in one than more methods got there own. A (bad) example for this would be extractIdFromUrl.

Another nice advantage off this:
Because this design exposes most of the internals the class can now also be used for other purposes than only being a download. It's now possible, for example, to log an instance in and use extractAllCourses to retrieve a list of all courses for other shenanigans the user may have in mind.
I didn't really plan for this use case, but I at least didn't want to rule it out :)

Btw, thanks for the quick answer

GityUpNow commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -73,6 +73,26 @@ def createIliasUrl(self, iliasid):
 		+ "&cmdNode=vi&baseClass=ilrepositorygui"
 
 
+	def extractIdFromUrl(self, url):
+		"""
+		Extracts the iliasid from a given ilias url.
+
+		:param      url:  ilias url containing a ref id
+		:type       url:  str
+
+		:returns:	iliasid of given course or object
+		:rtype:     int
+
+		:raises:	ValueError:  when an url without a iliasid is given.
+		"""
+
+		match = re.search(r"ref_id=(\d+)", url)
+		if match == None:
+			raise ValueError("No iliasid in given URL")
+
+		return int(match.group(1))
+
+
```
Coded it like this so it'll be easier to test for me. But maybe I went a bit overboard with this one :sweat_smile: 
I think it'll mostly boil down to the question: Do you want short methods like createIliasUrl, extractIdFromUrl and addCourse to make methods easier to read, or do you want long all-in-one methods to cut down on code size.
It's your project, so I don't know what you'd prefer. Thought tbh because of createIliasUrl and scanCourses that you'd be more into the former design, maybe I parsed that wrong.

jhelgert replied:
Hm, my code design isn't really consistent and shouldn't serve as a rule tbh. It's still a quick-and-dirt solution in the most parts of the code. If someone is willed to write tests for all methods at a later point, I agree on your approach with short methods. But I still think an own method for this one and `createIliasUrl()` is overkill and both should be inlined into their respective places.

GityUpNow commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -127,6 +173,44 @@ def addCourses(self, *iliasids):
 			self.addCourse(iliasid)
 
 
+	def addAllUserCourses(self, excluded_ids=[]):
+		"""
+		Extracts the users subscribed courses and adds them to the course list.
+
+		:param      excluded_ids:  optional ilias ids to ignore
+		:type       excluded_ids:  list
+
+		:returns:   number of courses added
+		:rtype:     int
+		"""
+
+		num_added = 0
+		courses = self.extractUserCourses()
+
+		for course in courses:
+			iliasid = course["iliasid"]
+
+			if iliasid not in excluded_ids:
+				self.addCourse(iliasid)
+				num_added += 1
+
+		return num_added
```
Yeah, never was really happy with this method name either. Couldn't think of a better one myself.
addAllCurrentSemesterCourses sounds pretty good, but I think I'd shorten it to addAllSemesterCourses.
If we add a semester parameter, that points by default to the current one, we could make the method a bit more flexible.
This way the user still has the option to download other semesters if they want to.

jhelgert replied:
Good idea, I'd agree on something like

``` python
def addAllSemesterCourses(self, only_current_semester=True, exclude_ids=None):
    # ...
```

GityUpNow replied:
I don't think limiting it to the current semester is neccessary. 
I'd rather go for a string, by default only the current one, potentially another semester(so specify the semester name) and maybe a wildcard if you want, well, all of them.
Otherwise looks good :)

jhelgert replied:
And there's another quirk in Ilias: If English is set as the language in the general settings, the semester string 'HWS 2020' is not for all courses translated to 'WT 2020', the same holds for 'FSS 2020' and 'ST 2020', respectively. Please don't ask me why 😅.

I'd be happy with something like

``` python
def __init__(self):
    # ...
    self.current_semester_pattern = # ... determine the current semester via datetime.now()
    # ...
```

``` python 
def addAllSemesterCourses(self, semester_pattern=None, exclude_ids=None):
    if semester_pattern:
        # ....
    else:
        # use self.current_semester_pattern
```   

where `current_semester_pattern` and `semester_pattern` are regex patterns. I guess then it might be a good idea to also add  1-2 further examples in the readme.md.

Commit:
Fix crash with Non-Video files in MediaCaptions

Co-authored-by: Jonathan Helgert <jhelgert@users.noreply.github.com>

GityUpNow commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -100,6 +120,29 @@ def login(self, login_id, login_pw):
 			raise ConnectionError("Couldn't log into ILIAS. Make sure your provided uni-id and the password are correct.")
 
 
+	def extractUserCourses(self):
+		"""
+		Extracts the users subscribed courses from the login_soup
+
+		:returns:   List of dicts, with keys name, url and iliasid
+		:rtype:     list
+		"""
+
+		courses = []
+
+		containers = self.login_soup.select("div.il_ContainerListItem")
+		for container in containers:
+			title_elem = container.select("a.il_ContainerItemTitle")[0]
+
+			course_name = self.cleanCourseName(title_elem.string)
+			iliasid = self.extractIdFromUrl(title_elem["href"])
+
+			url = self.createIliasUrl(iliasid)
+			courses += [{"name": course_name, "url": url, "iliasid": iliasid}]
+
+		return courses
```
I'd prefer the approach with the two methods. I think the tasks are logically distinct and will be a lot easier to write tests for this way.
See my design philosophy at the bottom if you're interested why I coded it this way.

GityUpNow replied:
>     * It adds **all** courses from the user's Ilias main page. Depending on the user's Ilias settings, even courses from past semesters.
Yeah, that's very suboptimal. I should probably redo that.
>     * In my case it adds also no regular semester courses like 'License Information (Student University of Mannheim)' or 'Digitalisierung in der Lehre (Hilfskräfte und Tutorinnen und Tutoren)'. I don't see the need to scan these folders.
How can we filter them out? Do all non regular courses miss the semester number, or do we need to find another way?
> A better solution might be to scan only the current semester's courses.
How about scanning by default only the current semester, but being able to download other semesters if wanted?

jhelgert replied:
> I'd prefer the approach with the two methods. I think the tasks are logically distinct and will be a lot easier to write tests for this way.

I see your point. I never indented to write tests for this small project (probably because it's my last semester),

> How about scanning by default only the current semester, but being able to download other semesters if wanted?

👍 

> How can we filter them out? Do all non regular courses miss the semester number, or do we need to find another way?

Yeah, I think checking the `il_ContainerItemTitle` for the regex pattern `r'([A-Z]{3} \d{4})'` will do it.

GityUpNow commented on IliasDownloaderUniMA/__init__.py:
```diff
@@ -110,9 +153,12 @@ def addCourse(self, iliasid):
-		course_name = soup.find_all("ol", "breadcrumb")[0].find_all('li')[2].get_text()
-		course_name = re.sub(r"\[.*\] ", "", course_name)
-		self.courses += [{'name' : course_name, 'url': url}]
+		breadcrumb = soup.find_all("ol", "breadcrumb")[0]
+		# Course name is part of course link
+		course_elem = breadcrumb.find_all("a", href=re.compile("ref_id=" + str(iliasid)))[0]
+
+		course_name = self.cleanCourseName(course_elem.get_text())
+		self.courses += [{'name': course_name, 'url': url, 'iliasid': iliasid }]
```
Finally one consistent thing it Ilias :smiley: 
I'll probably redo this with the semester parsing.

jhelgert commented:
TL;DR:  

- Fixing `extractUserCourses()` should be the highest prio.
- My code design isn't really consistent and shouldn't serve as a rule. 
- (New) Rule of thumb: Once used one-liners shouldn't be an own method (`createIliasUrl()` & `extractIliasId()`).
- If someone's willed to write tests, I'd agree on the refactoring into smaller methods as long as it isn't overkill. 
- I guess a good point to start might be `scanFolder()` where the two loops could be split into something like `scanVideos()` and `scanForSubFolders()`.
- Most of the past bugs were due to `_parseFileProperties()`, so refactoring (or at least commenting) this one might be a good idea.

jhelgert commented:
Okay, lazy sunday, so I cleaned up the code, added a few tests and fixed #23 and #24. The new version is already on pip. If you still want to contribute, you can focus on `addAllSemesterCourses()`. 

PS: Just ping me back in case you still need the hacktoberfest-accepted label :).

GityUpNow commented:
Sure, I'll focus on fixing that feature.
It'll probably be a bit messy to port that over.

Also, I'd still really like the hacktoberfest-accepted label ^^
The system will only count it after the pull request is merged, so I won't get it too soon anyway. Having it now would give me some piece of mind ^^

jhelgert commented:
Okay, lazy Tuesday morning (the last one for a while), so I implemented a first working version of `addAllSemesterCourses()` and added some tests, you can find it in the [0.5.0 branch](https://github.com/jhelgert/IliasDownloaderUniMA/commit/fdd523e59fdecd90bfb2f8a8a2edb5dca853e848). 

I hope you don't mind since it's been your idea 🙈. I guess there a still some tests needed for all methods. Are you willed to focus on the tests? Would really appreciate it! And sorry for the mess!

GityUpNow commented:
Oh, I actually really mind. 

Working on this project has been a mess.
Just for reference: I only published 50% of all the code I wrote, the rest was too redundant. I deleted it too not over clutter the project. But that was fine, I wanted to give this my all, to write the best code I can.
I wanted to be as civil as possible.

Then you fixed #23 and #24 yourself, without telling me first.
Okay, now 80% of my code was thrown away.
I swallowed the frustration, because it wouldn't have been productive, I couldn't tell how you'd react, plus I really wanted to contribute something to this project.

Now 100% of my work got thrown away, you implemented the tests I talked about and now even implemented my own feature, after you told me "how about you focus on that?".
And you did all of this without asking me beforehand. 
I didn't port the code to your new dev version yet, but what if I had started and completed it locally. In what situation would I have been? At that point 110% of my code would have been thrown away.

So yeah, I'm fuming. You can write those tests yourself, I'm not willing to dunkywork after you've trashed all of my code, after you've reaped all the fruits of my labour.
Have a nice life, I don't think I'll contribute to this project any time soon.

jhelgert commented:
You have the right to be annoyed. Communication is the key, and concerning this, I should have done much better. So I'm sorry about that. I want to comment on some points: 

> Then you fixed #23 and #24 yourself, without telling me first. [...]
> I swallowed the frustration because it wouldn't have been productive, I couldn't tell how you'd react, plus I really wanted to contribute something to this project.

It would be best if you told me. How should I have known? I got the impression I did you a favour, and you're okay with that as long as you'd get your hacktoberfest-accepted label. I can't see anything wrong with fixing bugs inside parts of my code.

> Now 100% of my work got thrown away

Closing this PR throws away all of my work on reviewing this PR.

> I didn't port the code to your new dev version yet, but what if I had started and completed it locally. In what situation would I have been? At that point, 110% of my code would have been thrown away.

In this case, I would have rejected my code to merge your code so that your work wasn't for nothing.

> So yeah, I'm fuming. [...] after you've reaped all the fruits of my labour.

Tbh, I don't see where I reaped your fruits of labour. It's been your idea, but the code and `addAllSemesterCourses()` especially, was written by myself. Still, I intended to accept your PR for the hacktoberfest as a reward for your work.

GityUpNow commented:
> You have the right to be annoyed. Communication is the key, and concerning this, I should have done much better. So I'm sorry about that. I want to comment on some points:
> 
> > Then you fixed #23 and #24 yourself, without telling me first. [...]
> > I swallowed the frustration because it wouldn't have been productive, I couldn't tell how you'd react, plus I really wanted to contribute something to this project.
> 
> It would be best if you told me. How should I have known? I got the impression I did you a favour, and you're okay with that as long as you'd get your hacktoberfest-accepted label. I can't see anything wrong with fixing bugs inside parts of my code.
>
Sorry, I didn't expect that we had any miscommunication there.
It would have been completely okay if you had asked first. 
But this way I had to watch as the code I poured hours in slowly got more and more redundant. 
Yes, there is nothing wrong with fixing bugs in your code normally. 
But in this case I already wrote fixes for those bugs. I tried to do them as good as I can and to create a commit history as clean as possible. This lead to me taking at least 3 times as long, as I could have normally done it.
Asking me if you could replace them with your own fixes would have been 100% okay. But not asking me and simply redoing it communicated that you didn't value my work all that much.
As I said, I originally didn't want to cause a fuss. I had to fight this Hacktober against my depression, so this meant that I had extremely limited amounts of energy. Putting more energy and starting an argument would have meant that simply no energy would have been left for other projects. (Remember, I'd have needed four PRs)
The other aspects: You're the project owner, that gives you a position of power. If you had rejected it for any reason, there would have been nothing that I could do against. I couldn't tell beforehand how you'd react to me telling you: "Please don't do that, why didn't you tell me anything? Throw away your changes!". 
I don't know what type of person you are and you pushed over my changes without asking me first, so I couldn't tell if you'd be willing to compromise, or if I'd catch you on the wrong foot, potentially enrage you and risk getting my PR closed.
So I was basically hoping that I could get this swiftly done and that you'd keep your promise that I could write addAllSemesterCourses.
What will on your side to fix this mess, in an acceptable way for both of us could I have assumed, after you broke your word and made my code redundant again?
> > Now 100% of my work got thrown away
> 
> Closing this PR throws away all of my work on reviewing this PR.
> 
We're both in a loose-loose situation here.
Yes, it sucks for you to throw away your work on this PR.
But I invested at least as much work into this, probably more.
I had hopes about this, I wanted to give it my all and finally contribute something good to a project. Writing code that would have been installable with a package manager would have been a small rite of passage for me tbh.
But these doors have closed for me now. See after the next quote for the continuation.
> > I didn't port the code to your new dev version yet, but what if I had started and completed it locally. In what situation would I have been? At that point, 110% of my code would have been thrown away.
> 
> In this case, I would have rejected my code to merge your code so that your work wasn't for nothing.
> 
How should I have known that beforehand? 
Also: This would have given not much room to not look like a whiny brat, that can't go with time and wants to unreasonably insist on getting their own version merged.
Getting my version instead of yours merged would now only feel shallow, like I'm a bad loser that can't swallow my anger.
It wouldn't be an accomplishment anymore for me, it would only make me feel guilty and and give me a bad conscience.
> > So yeah, I'm fuming. [...] after you've reaped all the fruits of my labour.
>
> Tbh, I don't see where I reaped your fruits of labour. It's been your idea, but the code and addAllSemesterCourses() especially, was written by myself. Still, I intended to accept your PR for the hacktoberfest as a reward for your work.
This is the fourth time now, at least, that exactly this happened.
Me having an idea, sharing it with people, sometimes even when I'm decently into a project. They then taking the idea, thinking "this sounds pretty cool, I'll do this myself", ignoring my work, doing it them self, often even using some of the ideas I had and getting all the credit for it. Saying "I had the idea!" to others doesn't really amount to much, it often just looks like me wanting a piece of the cake that I don't really deserve.
All of these incidents haven't been done with bad intentions btw. They either didn't realize how not okay that was, after I've told them multiple times, or simply thought that it wouldn't hurt me. And sometimes not even think about me at all.
You know why I didn't point out how much you ignoring my fixes for #23 and #24 and redoing it yourself hurt? Because I tried that in the past. Multiple times. It always got ignored. I always put even more effort into fixing such situations, without getting anything in return. So I just wanted to get this over before anything similar would happen again.
So yeah, no, I actually put work into this. Planning and intellectual work is work as well. I had the idea, worked it out, tracked down the bugs, pointed out where your code broke, wrote my own fixes and my own code. The former four are inseparably linked to the latter too. I didn't submit a bug report, I didn't submit a feature request. I submitted three bug fixes and a feature. 
You reimplementing my spadework without asking me first doesn't mean that you just reimplemented my ideas. It communicated that you don't value my work that much and wanted a fun challenge for yourself.
I know that you had the best intentions. But having the best intentions doesn't mean that you knew better than others what they want or need.

I hope this whole text doesn't sound like an accusation, I mostly wanted to explain my reasoning here, why I reacted certain ways and what behaviours on your part weren't okay.
There's sadly no option for me to close this situation in which I won't loose. I throw away all my code, but I at least don't have a merged PR that only feels like a failing on my side, not an achievement, but an proof for my incompetence.
I'm sorry that this throws away your work on this PR, I really am. But this is sadly the way that'll give me the least scars.
You don't have to worry about the hacktoberfest anymore btw, I dropped out yesterday. I wouldn't have made it in time anymore.

GityUpNow commented:
> 
> 
> You have the right to be annoyed. Communication is the key, and concerning this, I should have done much better. So I'm sorry about that. I want to comment on some points:

Thank you for the apology, I'm really thankful. Later in the message you'll probably understand a lot better, why I'm so thankful.
My communication was pretty bad as well, I got carried away a bit too much my emotions. I detailed you why I reacted certain ways below, hopefully we'll be able to understand each other a bit better.
We can shift this discussion over to a proper Messenger if you want. I feel like we're drifting further and further from technical discussion and more into trying to understand what went wrong.
> 
> > Then you fixed #23 and #24 yourself, without telling me first. [...]
> > I swallowed the frustration because it wouldn't have been productive, I couldn't tell how you'd react, plus I really wanted to contribute something to this project.
> 
> It would be best if you told me. How should I have known? I got the impression I did you a favour, and you're okay with that as long as you'd get your hacktoberfest-accepted label. I can't see anything wrong with fixing bugs inside parts of my code.

To be fair, yes, I should really have communicated that better. A little bit of context: I had project and project ideas, that I really cared about, stolen from me at least three times. The harder I think back these occurrences I remember. Whenever I confronted people with that I got ignored. So I really got conditioned that in such or similar situations talking back doesn't work, that my number one priority should be to finish or get out of projects as quickly as possible, to minimize the amount of mental scars that I get. I'm currently trying to get rid of this behavior, but it's really not easy. So yes, I really should have communicated that better and I deeply apologized that I reacted so instinctively. In my mind, pointing the situation out would only have made things worse. In retrospective I have to say that I was completely wrong there. You really couldn't know that.
That was my part of miscommunication.
I know now that you had the best intentions with closing #23 and #24 yourself. I at least know that afterwards. But in the situation I really couldn't know that. I submitted a PR with one feature and three bug fixes. You then wrote your own fixes to two of them without telling me first. What this communicated was that you didn't value my work(important, communicated. I don't want to accuse you of this, this is simply what it communicated to me) and would much rather fix these issues yourself. Yes, fixing bugs in your own code is completely okay. Normally. I didn't submit three bug reports, I submitted three bug reports with my own fixes. Seeing my fixes and then writing your own, without communicating that first, was perceived by me as simply throwing away my work.
At this point I really couldn't tell what type of person you are. You trashed two of my fixes without telling me first, so I really had to fear that you could be the type of person that would get angry when pointing such transgressions out. I honestly feared that pointing that out could get my PR closed, at which point, what was I supposed to do? I'd be pretty powerless in that situation.
Afterwards I of cause know that you had the best intentions. But the problem is: "Gut gemeint ist nicht gut gemacht."(Well-intentioned isn't necessarily well done).  
I don't want to get too close to you there, but the best way to hurt people is to do what you think is best for them, without asking them first. What we think people want is almost always radically different from what they actually want.
> > Now 100% of my work got thrown away
> 
> Closing this PR throws away all of my work on reviewing this PR.
> 

I reacted a bit too emotional to this. As I said, whenever I tried to find a compromise in the past, things just got worse for me. 
We can try to merge it again, but I think we'd have to talk more beforehand.
> > I didn't port the code to your new dev version yet, but what if I had started and completed it locally. In what situation would I have been? At that point, 110% of my code would have been thrown away.
> 
> In this case, I would have rejected my code to merge your code so that your work wasn't for nothing.
> 

I'm sorry, but I really couldn't have known that. I detailed what your behavior communicated to me(as I said, I don't want to accuse you, that this was your intention. It's simply what this communicated to me), but you made me the promise that I could code addAllSemesterCourses. After you broke your word there any hope I had for compromise was simply out of the window. 
Since having to throw away own work is highly enraging for most people, myself included, I really didn't think that you'd agree to that. Still, I should have at least asked first. We really had a lot of miscommunication here.
> > So yeah, I'm fuming. [...] after you've reaped all the fruits of my labour.
> 
> Tbh, I don't see where I reaped your fruits of labour. It's been your idea, but the code and `addAllSemesterCourses()` especially,
was written by myself. Still, I intended to accept your PR for the hacktoberfest as a reward for your work.

Stolen is a bit too harsh, I agree, but there sadly aren't any less harsh words that describe the same situation, so let me elaborate, why I called it that:
I had the idea, worked out some of the kinks, tracked down bugs and analyzed them etc. 
I didn't propose a feature. I wrote a PR with a feature. 
Coding the feature would have been perfectly fine with a feature request. 
But coding the feature yourself in the case of a PR feels very disrespectful and rude. The submitter didn't want to submit a bug report or feature request, they wanted to integrate actual code into the project. Simply talking the idea, without asking first, communicates that the maintainer doesn't value the work and simply wanted the idea.
The reason why I react to this so harshly ties back to what I mentioned earlier.
It happened to me at least three times now, that I shared project ideas with people, or showed them my current progress on projects and they simply took my ideas, ran with them and did the projects them self, often forcing my to throw away my own project. 
When I tried to do anything against that, I simply got ignored and at one occasion even guilt tripped. This left deep scars on me.
So yeah, although the definition of intellectual property can be a bit sketchy at times, intellectual property is still intellectual property. For example: If I had written a book with an interesting underlying story and you decided to write your own book, following the same narrative, without asking me first, you of cause wouldn't have stolen my book, but still the idea.
So to summarize it: I had the idea and did a lot of planning work, then wrote my code. You took idea and planning and wrote your own code without asking me. Since idea+planning+work are inherently linked imo taking 2/3 of that would still be stealing.

Things got a bit longer, I apologize for that. I mostly wanted to communicate why I reacted in certain ways and how the miscommunication between us happened. I hope nothing of this sounded like an accusation, although the last paragraph probably sounded a bit like that. I tried to keep it as neutral as possible, but had I formulated the last paragraph less offensive and more defensive I wouldn't have been able to communicate the problem.
If you have any more questions in general or regarding my answers we should probably switch to a proper messenger, I can offer you Telegram, Matrix, Signal and WhatsApp (although I really dislike that one).

Btw you don't have to worry about hacktoberfest anymore, I had to drop out.

jhelgert commented:
Let's discuss [here](https://discord.gg/SujByUWF)

GityUpNow commented:
👍

Commit:
Reverted all changes thus far

Specific ones will be readded. Mv instead of git mv was used,
this breaks many git features.

Fixing each commit manually is not worth it, so I chose this
approach.

Commit:
Merge remote-tracking branch 'upstream/0.5.x' into addAllCourses

Commit:
Rolled back addAllSemesterCourses to Docstring

Commit:
Added Semester Constants

Commit:
Replaced strings with consts in getCurrentSemester

Commit:
Wrote addAllSemesterCourses.

Saddly ended up as a slightly more efficient version of jhelgert code.

This way the most code can be reused and review will be easier.

Commit:
Fixed tests so pattern is used as parameter

Commit:
Fixed faulty test in test_addAllSemesterCourses

test_own_pattern_with_excluded_id asserted there to be four courses.
The "Digitalisierung in der Lehre" course is ignored and
"OPM 601 Supply Chain Management" excluded.
Therefore => "assert len(m.courses) == 3"

Commit:
Completed addAllSemesterCourses

Commit:
Use instance own current semester pattern

Commit:
Reverted getCurrentSemester to remove complexity

Commit:
Updated semester times with Uni MA website

See here: https://www.uni-mannheim.de/studium/termine/semesterzeiten

Semesters seem to start in February
